### ISSUE TITLE
Possible bug: SqlDataReader.ReadAsync hangs and produces high garbage collection.

### ISSUE THREAD

**Issue by aliveli186** [dotnet/SqlClient#439]
Tested on: System.Data.SqlClient versions "4.4.0", "4.8.0" and "4.8.1"

I'm trying to read from a table which has 8 records.
In ReadAsync it fetches the first 6 records quickly and hangs around 45 seconds to fetch the rest 2.
During that 45 seconds time produces very high garbage collection.

The synchronous version on the other hand fetches the records immediately and there is no visible garbage collection.

```
while (await reader.ReadAsync(cancellationToken).ConfigureAwait(false))
{
...
}
```

vs.

```
while (reader.Read())
{
...
}
```

**Comment by cheenamalhotra** [comment]
Thanks for the repro @aliveli186 .

The new home for SqlClient driver issues is "dotnet/sqlclient" repository, we will investigate and get back to you!

**Comment by Wraith2** [comment]
@aliveli186 we'll need the sproc and any other database definitions it depends on.

**Comment by aliveli186** [comment]
@Wraith2 The table it selects is a View which depends on few other tables, few joins. Can't provide the db schema now.

```
CREATE PROCEDURE [Logging].[ErrorLogSelect] (@Expressions nvarchar(MAX) = NULL)
AS
BEGIN
    SET NOCOUNT ON;

    DECLARE @S nvarchar(MAX) = N'
    SELECT *
      FROM [Logging].[ErrorLog]';

    IF (@Expressions IS NOT NULL)
        SELECT @S = @S + N'
		   '        + @Expressions;

    PRINT @S;
    PRINT 'EXECUTING...';

    EXECUTE [sp_executesql] @S;

END;

```

**Comment by tarurar** [comment]
@cheenamalhotra does the investigation clarified the issue? I am having same, `await reader.ReadAsync()` hangs forever.

**Comment by Wraith2** [comment]
Investigation stalled because there isn't a complete repo, see "Can't provide the db schema now." above. If you have a self contained repo we can run then it can continue. Looking back at the report though It smells like long async strings to me and the answer is to not do that, fetch the string sync or use a textreader stream.

**Comment by tarurar** [comment]
@Wraith2 yep, fetching sync solves the issue. And the string was indeed huge. about 260 Mb.

**Comment by Wraith2** [comment]
That's huge. Unless you actually need that as a string I strongly suggest using the GetTextReader overload in sequential mode, it'll save you a lot of GC work.

**Comment by roji** [comment]
Another variant of #593, possibly a dup?

**Comment by Wraith2** [comment]
Yup

**Comment by roji** [comment]
@cheenamalhotra consider closing as a dup of #593

**Comment by cheenamalhotra** [comment]
Sure! Thanks @roji and @Wraith2 :)
Closed as duplicate of #593

**Comment by Wraith2** [comment]
I do sort of have a plan on how to deal with the long strings but it requires a lot of changes throughout the library. 

The main problem is that each received packet starts the parse process again which we could avoid if we knew that we could just pick up from the end of the last packet's parse stall, but we can't do that because we can't tell a stall from a failure because all the TryRead* functions return Boolean where true is succeeded and failure could mean error or stall. If we replace those bools with OperationResult and flow that all the way through we could snapshot on stall and then resume from that snapshot at the next packet receipt.
